components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOArtifactResolver.java

```diff
@@ -44,53 +60,175 @@
+            log.warn("Artifact validity period (" + artifactInfo.getExpTimestamp() + ") has been " +
+                    "exceeded for artifact: " + artifactResolve.getArtifact().getArtifact());
+            isValid = false;
+        }
+
+        // Checking for issuer.
+        if (StringUtils.equals(artifactInfo.getAuthnReqDTO().getIssuer(), artifactResolve.getIssuer().getValue())) {
+
+            String tenantDomain = SAMLSSOUtil.getTenantDomainFromThreadLocal();
+            SAMLSSOServiceProviderDO serviceProviderDO = SAMLSSOUtil.getSPConfig(
+                    tenantDomain, SAMLSSOUtil.splitAppendedTenantDomain(artifactResolve.getIssuer().getValue()));
+
+            // Checking for signature.
+            if (serviceProviderDO.isDoValidateSignatureInArtifactResolve()) {
+                isValid = validateArtifactResolveSignature(artifactResolve, serviceProviderDO);
+            }
+        } else {
+            log.warn("Artifact Resolve Issuer: " + artifactResolve.getIssuer().getValue() + " is not valid.");
+            isValid = false;
+        }
+
+        return isValid;
+    }
+
+    /**
+     * Validate the signature of SAML2 artifact resolve object.
+     *
+     * @param artifactResolve   Artifact resolve object.
+     * @param serviceProviderDO Service provider object.
+     * @throws ArtifactBindingException
+     */
+    private boolean validateArtifactResolveSignature(ArtifactResolve artifactResolve,
+                                                  SAMLSSOServiceProviderDO serviceProviderDO)
+            throws ArtifactBindingException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Validating Artifact Resolve signature for artifact: " +
+                    artifactResolve.getArtifact().getArtifact() + ", issuer: " +
+                    artifactResolve.getIssuer().getValue());
+        }
+
+        if (artifactResolve.getSignature() == null) {
+            log.warn("Signature was null in SAML2 Artifact Resolve with artifact: " +
+                    artifactResolve.getArtifact().getArtifact() + " issuer: " + artifactResolve.getIssuer().getValue());
+            return false;
+        }
+        SignatureImpl signImpl = (SignatureImpl) artifactResolve.getSignature();
+
+        if (serviceProviderDO.getX509Certificate() == null) {
+            throw new ArtifactBindingException("Artifact resolve signature validation is enabled, but SP " +
+                    "doesn't have a certificate");
+        }
+
+        try {
+            BasicX509Credential credential = new BasicX509Credential();
+            credential.setEntityCertificate(serviceProviderDO.getX509Certificate());
+            SignatureValidator validator = new SignatureValidator(credential);
+            validator.validate(signImpl);
+            return true;
+        } catch (ValidationException e) {
+            log.warn("Signature validation failed for SAML2 Artifact Resolve with artifact: " +
```

Comment:
log 'e' also. may be as a separate debug log if this is expected to be logged too often